File: code/regex/nfa.py
# Regex -> NFA (Sedgewick's one-state-per-character encoding) + set-of-states
# simulation. Syntax: letters, concatenation, | (one per group), *, ( ).
class NFA:
    def __init__(self, pattern):
        self.re = "(" + pattern + ")"   # implicit wrap anchors top-level | or *
        self.m = len(self.re)           # state m is the accept state
        self.eps = [[] for _ in range(self.m + 1)]
        ops = []                        # stack of '(' and '|' positions
        for i, c in enumerate(self.re):
            lp = i                      # left end of the atom/group ending at i
            if c in "(|":
                ops.append(i)
            elif c == ")":
                orpos = ops.pop()
                if self.re[orpos] == "|":          # wire the alternation
                    lp = ops.pop()
                    self.eps[lp].append(orpos + 1) # '(' skips into branch 2
                    self.eps[orpos].append(i)      # '|' skips to the ')'
                else:
                    lp = orpos                     # plain group: lp = the '('
            if i < self.m - 1 and self.re[i + 1] == "*":   # closure lp..i
                self.eps[lp].append(i + 1)         # skip the operand
                self.eps[i + 1].append(lp)         # repeat the operand
            if c in "(*)":
                self.eps[i].append(i + 1)          # fall through the meta char

    def closure(self, s):
        """e-closure(S): DFS over the e-digraph -- all states reachable free."""
        seen, stack = set(s), list(s)
        while stack:
            u = stack.pop()
            for v in self.eps[u]:
                if v not in seen:
                    seen.add(v)
                    stack.append(v)
        return seen

    def matches(self, text):
        """Does the NFA accept text? Track the reachable-state set: O(mn)."""
        pc = self.closure({0})                     # states reachable on ""
        for c in text:
            nxt = {u + 1 for u in pc               # advance every letter state
                   if u < self.m and self.re[u] == c}
            pc = self.closure(nxt)                 # then close over e-edges
            if not pc:
                return False                       # dead: nothing can revive
        return self.m in pc                        # accept state in final set?

File: code/regex/nfa.java
// Regex -> NFA (Sedgewick's one-state-per-character encoding) + set-of-states
// simulation. Syntax: letters, concatenation, | (one per group), *, ( ).
class NFA {
    final String re;              // wrapped regex: "(" + pattern + ")"
    final int m;                  // re.length(); state m is the accept state
    final List<List<Integer>> eps = new ArrayList<>();

    NFA(String pattern) {
        re = "(" + pattern + ")"; // implicit wrap anchors a top-level | or *
        m = re.length();
        for (int i = 0; i <= m; i++) eps.add(new ArrayList<>());
        Deque<Integer> ops = new ArrayDeque<>();  // '(' and '|' positions
        for (int i = 0; i < m; i++) {
            int lp = i;           // left end of the atom/group ending at i
            char c = re.charAt(i);
            if (c == '(' || c == '|') ops.push(i);
            else if (c == ')') {
                int orPos = ops.pop();
                if (re.charAt(orPos) == '|') {    // wire the alternation
                    lp = ops.pop();
                    eps.get(lp).add(orPos + 1);   // '(' skips into branch 2
                    eps.get(orPos).add(i);        // '|' skips to the ')'
                } else lp = orPos;                // plain group: lp = the '('
            }
            if (i < m - 1 && re.charAt(i + 1) == '*') {  // closure lp..i
                eps.get(lp).add(i + 1);           //   skip the operand
                eps.get(i + 1).add(lp);           //   repeat the operand
            }
            if (c == '(' || c == '*' || c == ')')
                eps.get(i).add(i + 1);            // fall through the meta char
        }
    }

    // e-closure(S): DFS over the e-digraph -- all states reachable for free
    Set<Integer> closure(Set<Integer> s) {
        Set<Integer> seen = new TreeSet<>(s);
        Deque<Integer> stack = new ArrayDeque<>(s);
        while (!stack.isEmpty()) {
            int u = stack.pop();
            for (int v : eps.get(u))
                if (seen.add(v)) stack.push(v);
        }
        return seen;
    }

    // does the NFA accept text? -- track the reachable-state set: O(mn)
    boolean matches(String text) {
        Set<Integer> pc = closure(new TreeSet<>(Set.of(0)));
        for (char c : text.toCharArray()) {
            Set<Integer> next = new TreeSet<>();
            for (int u : pc)                      // advance every letter state
                if (u < m && re.charAt(u) == c) next.add(u + 1);
            pc = closure(next);                   // then close over e-edges
            if (pc.isEmpty()) return false;       // dead: nothing can revive
        }
        return pc.contains(m);                    // accept state in final set?
    }
}
